synthesizer/src/snark/batch/mod.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use anyhow::anyhow;
+
+use super::*;
+
+mod bytes;
+mod parse;
+mod serialize;
+
+use std::collections::BTreeMap;
+
+#[derive(Clone, PartialEq, Eq)]
+pub enum Key<N: Network> {
+    ProvingKey(ProvingKey<N>),
+    VerifyingKey(VerifyingKey<N>),
+}
+
+#[derive(Clone)]
+pub enum KeyMode {
+    Proving = 0,
+    Verifying = 1,
+}
+
+
+#[derive(Clone)]
+pub struct KeyBatch<N: Network> {
+    /// A batch of proving or verifying keys
+    batch: Vec<Key<N>>,
+    mode: KeyMode,
+}
+
+impl<N: Network> KeyBatch<N> {
+    /// Initializes a new batch.
+    pub(crate) fn new(
+        keys: usize,
+        mode: KeyMode,
+    ) -> Self {
+        Self { batch: Vec::with_capacity(keys), mode }
+    }
+
+    /// Adds keys to the batch.
+    pub(crate) fn add(
+        &mut self,
+        key: Key<N>,
+    ) -> Result<(), anyhow::Error> {
+        // TODO: consider creating custom error type. We use error() in this folder elsewhere
+        match key {
+            Key::VerifyingKey(_) => if matches!(self.mode,  KeyMode::Proving) {
+                return Err(anyhow!("Cannot add a verifying key to a proving batch"))
```

Comment:
```suggestion
                bail!("Cannot add a verifying key to a proving batch");
```
nit; applies to other spots like this

Reply:
Done